File: 02a/python/main.py
from collections import defaultdict


with open('input.txt', 'rt') as f:
    lines = (line.strip() for line in f.readlines())

d = defaultdict(lambda: 0)

def parse(line):
    direction, amt = line.split()
    amt = int(amt)
    d[direction] += amt

# Force execution of the generator.
[_ for _ in map(parse, lines)]

result = (d['down'] - d['up']) * (d['forward'] - d['back'])
print(result)

File: 02a/ruby/main.rb
FILENAME = "input.txt"
file = File.open(FILENAME)
lines = file.readlines.map(&:chomp)

$horizontal = 0
$depth = 0

def parse_line(line)
  direction, qty = line.split
  qty = qty.to_i
  case direction
  when "forward"
    $horizontal += qty
  when "up"
    $depth -= qty
  when "down"
    $depth += qty
  else
    puts "invalid input"
  end
end

lines.each { |line| parse_line(line) }

puts $horizontal * $depth
